app/src/main/java/com/babylon/wallet/android/presentation/history/HistoryViewModel.kt

```diff
@@ -0,0 +1,420 @@
+                historyItems.add(HistoryItem.Date(date))
+                transactionItems.forEach { historyItems.add(HistoryItem.Transaction(it)) }
+            }
+            it.copy(
+                content = if (historyItems.isEmpty()) {
+                    Content.Empty
+                } else {
+                    Content.Loaded(
+                        historyData = historyData,
+                        historyItems = historyItems.toPersistentList()
+                    )
+                },
+                loadMoreState = if (resetLoadingMoreState) null else it.loadMoreState
+            )
+        }
+    }
+
+    companion object {
+        private const val SCROLL_DELAY_AFTER_LOAD = 300L
+    }
+}
+
+internal sealed interface HistoryEvent : OneOffEvent {
+    data class OnTransactionItemClick(val url: String) : HistoryEvent
+    data class ScrollToItem(val index: Int, val loadingOffset: Boolean = false) : HistoryEvent
+    data class ScrollToTimeFilter(val index: Int) : HistoryEvent
+}
+
+data class State(
+    val accountWithAssets: AccountWithAssets? = null,
+    val content: Content = Content.Loading,
+    val uiMessage: UiMessage? = null,
+    val showFiltersSheet: Boolean = false,
+    val filters: HistoryFilters = HistoryFilters(),
+    val timeFilterItems: ImmutableList<Selectable<MonthFilter>> = persistentListOf(),
+    val loadMoreState: LoadingMoreState? = null,
+    val firstVisibleIndex: Int? = null,
+    val lastVisibleIndex: Int? = null,
+    val ignoreScrollEvents: Boolean = false
+) : UiState {
+
+    val historyData
+        get() = when (content) {
+            is Content.Loaded -> content.historyData
+            else -> null
+        }
+
+    val shouldShowFiltersButton: Boolean
+        get() = content is Content.Loaded || filters.isAnyFilterSet
+
+    val shouldEnableUserInteraction: Boolean
+        get() = content !is Content.Loading && loadMoreState == null
+
+    val historyItems
+        get() = when (content) {
+            is Content.Loaded -> content.historyItems
+            else -> null
+        }
+
+    private val currentFilters: HistoryFilters?
+        get() = when (content) {
+            is Content.Loaded -> content.historyData.filters
+            else -> null
+        }
+
+    val fungibleResources: List<Resource.FungibleResource>
+        get() = accountWithAssets?.assets?.knownFungibles.orEmpty()
+
+    val nonFungibleResources: List<Resource.NonFungibleResource>
+        get() = accountWithAssets?.assets?.knownNonFungibles.orEmpty()
+
+    val canLoadMoreUp: Boolean
+        get() = content is Content.Loaded && loadMoreState == null && content.historyData.prevCursorId != null
+
+    val canLoadMoreDown: Boolean
+        get() = content is Content.Loaded && loadMoreState == null && content.historyData.nextCursorId != null
+
+    val filtersChanged: Boolean
+        get() = filters != currentFilters
+}
+
+sealed class Content {
```

Comment:
I still think these classes `Content`, `MonthFilter`, `LoadingMoreState`, ... should reside under the same namespace of the `State` data class.